cl/favorites/utils.py

```diff
@@ -200,30 +259,99 @@ def send_prayer_emails(instance: RECAPDocument) -> None:
 
 
 async def get_user_prayer_history(user: User) -> tuple[int, float]:
-    filtered_list = Prayer.objects.filter(user=user, status=Prayer.GRANTED)
+    filtered_list = Prayer.objects.filter(
+        user=user, status=Prayer.GRANTED
+    ).select_related("recap_document")
 
     count = await filtered_list.acount()
 
-    total_cost = 0
-    async for prayer in filtered_list:
-        total_cost += float(price(prayer.recap_document))
+    cost = await (
+        filtered_list.values("recap_document")
+        .distinct()
+        .annotate(
+            price=Case(
+                When(recap_document__is_free_on_pacer=True, then=Value(0.0)),
+                When(
+                    recap_document__page_count__gt=0,
+                    then=Least(
+                        Value(3.0),
+                        F("recap_document__page_count") * Value(0.10),
+                    ),
+                ),
+                default=Value(0.0),
+            )
+        )
+        .aaggregate(Sum("price", default=0.0))
+    )
```

Comment:
I think we can avoid the duplicate code in this method and in `get_lifetime_prayer_stats` by encapsulating the shared logic into a separate method, something like:

```
async def compute_cost(queryset):
    return await (
        queryset.values("recap_document")
        .distinct()
        .annotate(
            price=Case(
                When(recap_document__is_free_on_pacer=True, then=Value(0.0)),
                When(
                    recap_document__page_count__gt=0,
                    then=Least(
                        Value(3.0),
                        F("recap_document__page_count") * Value(0.10),
                    ),
                ),
                default=Value(0.0),
            )
        )
        .aaggregate(Sum("price", default=0.0))
    )
```

And then just using:
`cost =  await compute_cost(filtered_list)`

OR:

`total_cost = await compute_cost(prayer_by_status.select_related("recap_document"))
`

Reply:
You're correct! I'll create this new helper function.